File: frontend/summonerinsight-app/src/pages/SummonerPage/components/reusableFunctions.js
export function parseDescriptionToJson(description) {
    const result = {
        stats: {},
        passive: [],
        active: []
    };

    // Match stats section and extract each line
    const statsPattern = /<stats>(.*?)<\/stats>/s;
    const statsMatch = description.match(statsPattern);
    if (statsMatch) {
        const statsText = statsMatch[1];
        const statLines = statsText.split(/<br>/).filter(line => line.trim());

        statLines.forEach(line => {
            const attentionMatch = line.match(/<attention>(\d+%?)<\/attention>\s*(.+)/);
            if (attentionMatch) {
                const [_, value, text] = attentionMatch;
                // Capitalize the first letter of each key word
                const key = text.split(' ').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
                result.stats[key] = value.includes('%') ? value : Number(value);  // Handle % values as strings
            }
        });
    }

    // Match passive abilities and capture all occurrences
    const passivePattern = /<passive>(.*?)<\/passive>\s*<br>(.*?)(?=<br><br>|<passive>|<active>|$)/gs;
    let passiveMatch;
    while ((passiveMatch = passivePattern.exec(description)) !== null) {
        const name = passiveMatch[1].trim();
        let descriptionText = passiveMatch[2].trim();
        descriptionText = descriptionText.replace(/<[^>]+>/g, ''); // Remove any tags

        result.passive.push({
            name: name.charAt(0).toUpperCase() + name.slice(1).toLowerCase(),
            description: descriptionText
        });
    }

    // Match active abilities (if any)
    const activePattern = /<active>(.*?)<\/active>\s*<br>(.*?)(?=<br><br>|<active>|<passive>|$)/gs;
    let activeMatch;
    while ((activeMatch = activePattern.exec(description)) !== null) {
        const name = activeMatch[1].trim();
        let descriptionText = activeMatch[2].trim();
        descriptionText = descriptionText.replace(/<[^>]+>/g, ''); // Remove any tags

        result.active.push({
            name: name.charAt(0).toUpperCase() + name.slice(1).toLowerCase(),
            description: descriptionText
        });
    }

    return result;
}

export const find_positions = (matchHistory, summonerProfile) => {
    let positions = {
        "TOP": 0,
        "JUNGLE": 0,
        "MIDDLE": 0,
        "BOTTOM": 0,
        "UTILITY": 0
    };
    matchHistory.forEach(match => {
        const position = match.participants.find(participant => participant.gameName === summonerProfile.gameName).position;
        positions[position]++;
    });
    return Object.keys(positions).reduce((acc, key) => {
        if (positions[key] > 5) {
            acc[key] = positions[key];
        }
        return acc;
    }, {});
};

File: backend/ReusableFunctions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace backend
{
    public class ReusableFunctions
    {
        public JsonObject ParseDescriptionToJson(string description)
        {
            var result = new JsonObject();
            var stats = new JsonObject();
            var passives = new JsonArray();
            var actives = new JsonArray();

            // Match statistics
            var statMatches = Regex.Matches(description, @"<stats>(.*?)<\/stats>", RegexOptions.Singleline);
            foreach (Match statMatch in statMatches)
            {
                var statText = statMatch.Groups[1].Value;
                var statLines = statText.Split(new[] { "<br>" }, StringSplitOptions.RemoveEmptyEntries);
                
                foreach (var line in statLines)
                {
                    // Extract attention value and text (e.g., "110 Health")
                    var attentionMatch = Regex.Match(line, @"<attention>(?<value>[\d]+)<\/attention>\s*(?<text>.+)");
                    if (attentionMatch.Success)
                    {
                        string value = attentionMatch.Groups["value"].Value;
                        string text = attentionMatch.Groups["text"].Value.Trim();

                        // Normalize key (e.g., "health" from "Health")
                        string key = text.ToLower(); // Use the text as-is, may want to clean it up further
                        stats[key] = double.TryParse(value, out double numericValue) ? numericValue : value;
                    }
                }
            }

            // Match passive abilities
            var passivePattern = @"<passive>(?<name>.*?)<\/passive>\s*<br>(?<desc>.*?)(?=<br><br>|$)";
            var passiveMatches = Regex.Matches(description, passivePattern, RegexOptions.Singleline);

            foreach (Match passiveMatch in passiveMatches)
            {
                var name = passiveMatch.Groups["name"].Value.Trim();
                var descriptionText = passiveMatch.Groups["desc"].Value.Trim();

                // Remove HTML tags from descriptionText
                descriptionText = Regex.Replace(descriptionText, @"<[^>]*>", "");

                // Add the passive to the array
                passives.Add(new JsonObject
                {
                    ["name"] = name,
                    ["description"] = descriptionText
                });
            }

            // Match active abilities
            var activePattern = @"<active>(?<name>.*?)<\/active>\s*<br>(?<desc>.*?)(?=<br><br>|$)";
            var activeMatches = Regex.Matches(description, activePattern, RegexOptions.Singleline);

            foreach (Match activeMatch in activeMatches)
            {
                var name = activeMatch.Groups["name"].Value.Trim();
                var descriptionText = activeMatch.Groups["desc"].Value.Trim();

                // Remove HTML tags from descriptionText
                descriptionText = Regex.Replace(descriptionText, @"<[^>]*>", "");

                // Add the passive to the array
                actives.Add(new JsonObject
                {
                    ["name"] = name,
                    ["description"] = descriptionText
                });
            }

            // Assemble the final result
            result["stats"] = stats;
            result["passive"] = passives;
            result["active"] = actives;

            return result;
        }
    }
}